packages/babel-traverse/src/path/lib/hoister.js

```diff
@@ -2,11 +2,23 @@ import { react } from "babel-types";
 import * as t from "babel-types";
 
 const referenceVisitor = {
+  // This visitor looks for bindings to establish a topmost scope for hoisting.
   ReferencedIdentifier(path, state) {
-    if (path.isJSXIdentifier() && react.isCompatTag(path.node.name)) {
+    // Don't hoist regular JSX identifiers ('div', 'span', etc).
+    // We do have to consider member expressions for hoisting (e.g. `this.component`)
+    if (path.isJSXIdentifier() && react.isCompatTag(path.node.name) && !path.parentPath.isJSXMemberExpression()) {
       return;
     }
 
+    // If the identifier refers to `this`, we need to break on the closest non-arrow scope.
+    if (path.node.name === "this") {
```

Comment:
Oh gross, I didn't know JSX didn't use `ThisExpression`. I bet we have bugs if you transpile ES6 -> ES5 with JSX pass-through enabled. No way does everything that processes `this` also check for JSXIdentifier.

Nothing for you to change, just complaining :P